contrib/encrypt_files_in_dir_to_s3.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python2.7
+# Copyright (C) 2016 UCSC Computational Genomics Lab
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Author : Arjun Arkal Rao
+Affiliation : UCSC BME, UCSC Genomics Institute
+File : contrib/encrypt_files_in_dir_to_s3.py
+
+SOURCE: https://github.com/jvivian/one_off_scripts/blob/master/
+        encrypt_files_in_dir_to_s3.py
+ORIGINAL AUTHOR: John Vivian
+
+Move files in a directory, or entire directory structures to S3 with (or without) encryption.
+"""
+from __future__ import print_function
+import argparse
+import base64
+import hashlib
+import os
+import subprocess
+import sys
+import re
+
+from boto.s3.connection import S3Connection
+
+
+class InputParameterError(Exception):
+    """
+    This Error Class will be raised  in the case of a bad parameter provided.
+    """
+    __module__ = Exception.__module__
+
+
+def generate_unique_key(master_key, url):
+    """
+    This module will take a master key and a url, and then make a new key specific to the url, based
+    off the master.
+    :param str master_key: Path to the master key used for encryption.
+    :param str url: Full URL to the potential file location in S3.
+    :returns new_key: The new key that is obtained by hashing the master key:url combination.
+    """
+    with open(master_key, 'r') as keyfile:
+        master_key = keyfile.read()
+    assert len(master_key) == 32, 'Invalid Key! Must be 32 characters.  Key: %s' % master_key + \
+                                  ', Length: %s' % len(master_key)
+    new_key = hashlib.sha256(master_key + url).digest()
+    assert len(new_key) == 32, 'New key is invalid and is not ' + \
+        '32 characters: {}'.format(new_key)
+    return new_key
+
+
+def get_bucket_region(bucket_name, aws_creds):
+    """
+    Queries S3 to identify the region hosting the provided bucket.
+    :param str bucket_name: Name of an S3 bucket
+    :param dict aws_creds: Credentials to use for the query
+    :return: Region hosting the bucket
+    """
+    conn = S3Connection(aws_creds['aws_access_key_id'], aws_creds['aws_secret_access_key'])
+    bucket = conn.lookup(bucket_name)
+    if not bucket:
+        # TODO: Make the bucket here?
+        raise InputParameterError('The provided bucket %s doesn\'t exist' % bucket_name)
+    return str(bucket.get_location())
```

Comment:
So looking at [this](http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region),  the http endpoints for us-east-1 are `s3.amazonaws.com` and `s3-external-1.amazonaws.com` so i'm going to return `external-1`.